// @azure/storage-file-datalake/dist/commonjs/models.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageDataLakeAudience = exports.PathResourceType = exports.PathRenameMode = exports.PathGetPropertiesAction = void 0;
exports.getDataLakeServiceAccountAudience = getDataLakeServiceAccountAudience;
// Keeping these for backward compatibility when we changed to use string unions.
/**
 * Defines values for PathGetPropertiesAction.
 * Possible values include: 'getAccessControl', 'getStatus'
 * @readonly
 */
var PathGetPropertiesAction;
(function (PathGetPropertiesAction) {
    PathGetPropertiesAction["GetAccessControl"] = "getAccessControl";
    PathGetPropertiesAction["GetStatus"] = "getStatus";
})(PathGetPropertiesAction || (exports.PathGetPropertiesAction = PathGetPropertiesAction = {}));
/**
 * Defines values for PathRenameMode.
 * Possible values include: 'legacy', 'posix'
 * @readonly
 */
var PathRenameMode;
(function (PathRenameMode) {
    PathRenameMode["Legacy"] = "legacy";
    PathRenameMode["Posix"] = "posix";
})(PathRenameMode || (exports.PathRenameMode = PathRenameMode = {}));
/**
 * Defines values for PathResourceType.
 * Possible values include: 'directory', 'file'
 * @readonly
 */
var PathResourceType;
(function (PathResourceType) {
    PathResourceType["Directory"] = "directory";
    PathResourceType["File"] = "file";
})(PathResourceType || (exports.PathResourceType = PathResourceType = {}));
/**
 * Defines the known cloud audiences for Storage.
 */
var StorageDataLakeAudience;
(function (StorageDataLakeAudience) {
    /**
     * The OAuth scope to use to retrieve an AAD token for Azure Storage.
     */
    StorageDataLakeAudience["StorageOAuthScopes"] = "https://storage.azure.com/.default";
})(StorageDataLakeAudience || (exports.StorageDataLakeAudience = StorageDataLakeAudience = {}));
/**
 * To get OAuth audience for a storage account for datalake service.
 */
function getDataLakeServiceAccountAudience(storageAccountName) {
    return `https://${storageAccountName}.dfs.core.windows.net/.default`;
}
/** *********************************************************/
/** DataLakeLeaseClient option and response related models */
/** *********************************************************/
